File: test_tool_client.py
"""
Test script để mô phỏng tool Python lấy resources từ server
Run: python test_tool_client.py
"""

import requests
import json

# Server config
SERVER_URL = "http://localhost:3001"
USERNAME = "xzzz"  # Hoặc username bạn muốn test
PASSWORD = "123"   # Password của user đó

def login(username, password):
    """Login và lấy JWT token"""
    print(f"🔐 Đang login với user: {username}...")
    
    response = requests.post(
        f"{SERVER_URL}/api/auth/login",
        json={
            "username": username,
            "password": password
        }
    )
    
    if response.status_code == 200:
        data = response.json()
        token = data.get('token')
        user_id = data.get('userId')
        print(f"✅ Login thành công!")
        print(f"   User ID: {user_id}")
        print(f"   Token: {token[:20]}...{token[-10:]}\n")
        return token, user_id
    else:
        print(f"❌ Login thất bại: {response.status_code}")
        print(f"   Response: {response.text}")
        return None, None

def get_resources(user_id, token=None):
    """Lấy resources của user từ server"""
    print(f"📦 Đang lấy resources cho user ID: {user_id}...")
    
    headers = {}
    if token:
        headers['Authorization'] = f'Bearer {token}'
    
    response = requests.get(
        f"{SERVER_URL}/api/resources/{user_id}",
        headers=headers
    )
    
    if response.status_code == 200:
        data = response.json()
        
        api_keys = data.get('api_keys', [])
        proxies = data.get('proxies', [])
        rotating_keys = data.get('rotating_proxy_keys', [])
        
        print(f"✅ Lấy resources thành công!\n")
        
        # Display API Keys
        print(f"🔑 API Keys: {len(api_keys)} keys")
        for i, key in enumerate(api_keys[:3], 1):  # Show first 3
            api_key = key.get('api_key', '')
            print(f"   {i}. {api_key[:10]}...{api_key[-10:]}")
        if len(api_keys) > 3:
            print(f"   ... và {len(api_keys) - 3} keys nữa")
        
        # Display Proxies
        print(f"\n🌐 Proxies: {len(proxies)} proxies")
        for i, proxy in enumerate(proxies[:3], 1):  # Show first 3
            proxy_url = proxy.get('proxy_url', 'N/A')
            print(f"   {i}. {proxy_url[:50]}")
        if len(proxies) > 3:
            print(f"   ... và {len(proxies) - 3} proxies nữa")
        
        # Display Rotating Keys
        print(f"\n🔄 Rotating Proxy Keys: {len(rotating_keys)} keys")
        for i, key in enumerate(rotating_keys[:3], 1):  # Show first 3
            api_key = key.get('api_key', '')
            print(f"   {i}. {api_key[:10]}...{api_key[-10:]}")
        if len(rotating_keys) > 3:
            print(f"   ... và {len(rotating_keys) - 3} keys nữa")
        
        # Save to file
        output_file = f"resources_{USERNAME}.json"
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"\n💾 Đã lưu vào file: {output_file}")
        
        return data
    else:
        print(f"❌ Lỗi lấy resources: {response.status_code}")
        print(f"   Response: {response.text}")
        return None

def main():
    print("=" * 60)
    print("🧪 TEST TOOL CLIENT - FETCH RESOURCES FROM SERVER")
    print("=" * 60)
    print()
    
    # Step 1: Login
    token, user_id = login(USERNAME, PASSWORD)
    
    if not token or not user_id:
        print("\n❌ Không thể tiếp tục test do login thất bại")
        return
    
    print("-" * 60)
    
    # Step 2: Get Resources
    resources = get_resources(user_id, token)
    
    if resources:
        print("\n" + "=" * 60)
        print("✅ TEST THÀNH CÔNG!")
        print("=" * 60)
        print("\n📝 Workflow cho tool thật:")
        print("   1. User nhập username/password trong tool")
        print("   2. Tool gọi /api/auth/login → lấy token + userId")
        print("   3. Tool gọi /api/resources/{userId} → lấy tất cả resources")
        print("   4. Tool load API keys, proxies, rotating keys vào memory")
        print("   5. Tool sẵn sàng xử lý batch!")
    else:
        print("\n❌ TEST THẤT BẠI!")

if __name__ == "__main__":
    try:
        main()
    except requests.exceptions.ConnectionError:
        print("❌ Không thể kết nối tới server!")
        print("   Đảm bảo server đang chạy: npm run dev")
    except Exception as e:
        print(f"❌ Lỗi: {e}")

File: test-tool-client.js
/**
 * Test script để mô phỏng tool lấy resources từ server
 * Run: node test-tool-client.js
 */

const https = require('http');

// Config
const SERVER_URL = 'http://localhost:3001';
const USERNAME = 'xzzz';  // Thay bằng username của bạn  
const PASSWORD = 'password_cua_ban';   // Thay bằng password thật của user xzzz

async function apiRequest(method, path, data = null, token = null) {
  const url = new URL(path, SERVER_URL);
  
  const options = {
    method: method,
    headers: {
      'Content-Type': 'application/json',
    }
  };
  
  if (token) {
    options.headers['Authorization'] = `Bearer ${token}`;
  }
  
  return new Promise((resolve, reject) => {
    const req = https.request(url, options, (res) => {
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => {
        try {
          resolve({
            status: res.statusCode,
            data: JSON.parse(body)
          });
        } catch (e) {
          resolve({
            status: res.statusCode,
            data: body
          });
        }
      });
    });
    
    req.on('error', reject);
    if (data) req.write(JSON.stringify(data));
    req.end();
  });
}

async function login(username, password) {
  console.log(`🔐 Đang login với user: ${username}...`);
  
  const response = await apiRequest('POST', '/api/auth/login', {
    username,
    password
  });
  
  if (response.status === 200) {
    const { token, userId } = response.data;
    console.log(`✅ Login thành công!`);
    console.log(`   User ID: ${userId}`);
    console.log(`   Token: ${token.substring(0, 20)}...${token.slice(-10)}\n`);
    return { token, userId };
  } else {
    console.log(`❌ Login thất bại: ${response.status}`);
    console.log(`   Response:`, response.data);
    return null;
  }
}

async function getResources(userId, token) {
  console.log(`📦 Đang lấy resources cho user ID: ${userId}...`);
  
  const response = await apiRequest('GET', `/api/resources/${userId}`, null, token);
  
  if (response.status === 200) {
    const data = response.data;
    const apiKeys = data.api_keys || [];
    const proxies = data.proxies || [];
    const rotatingKeys = data.rotating_proxy_keys || [];
    
    console.log(`✅ Lấy resources thành công!\n`);
    
    // Display API Keys
    console.log(`🔑 API Keys: ${apiKeys.length} keys`);
    apiKeys.slice(0, 3).forEach((key, i) => {
      const apiKey = key.api_key || '';
      console.log(`   ${i + 1}. ${apiKey.substring(0, 10)}...${apiKey.slice(-10)}`);
    });
    if (apiKeys.length > 3) {
      console.log(`   ... và ${apiKeys.length - 3} keys nữa`);
    }
    
    // Display Proxies
    console.log(`\n🌐 Proxies: ${proxies.length} proxies`);
    proxies.slice(0, 3).forEach((proxy, i) => {
      const proxyUrl = proxy.proxy_url || 'N/A';
      console.log(`   ${i + 1}. ${proxyUrl.substring(0, 50)}`);
    });
    if (proxies.length > 3) {
      console.log(`   ... và ${proxies.length - 3} proxies nữa`);
    }
    
    // Display Rotating Keys
    console.log(`\n🔄 Rotating Proxy Keys: ${rotatingKeys.length} keys`);
    rotatingKeys.slice(0, 3).forEach((key, i) => {
      const apiKey = key.api_key || '';
      console.log(`   ${i + 1}. ${apiKey.substring(0, 10)}...${apiKey.slice(-10)}`);
    });
    if (rotatingKeys.length > 3) {
      console.log(`   ... và ${rotatingKeys.length - 3} keys nữa`);
    }
    
    // Save to file
    const fs = require('fs');
    const outputFile = `resources_${USERNAME}.json`;
    fs.writeFileSync(outputFile, JSON.stringify(data, null, 2));
    console.log(`\n💾 Đã lưu vào file: ${outputFile}`);
    
    return data;
  } else {
    console.log(`❌ Lỗi lấy resources: ${response.status}`);
    console.log(`   Response:`, response.data);
    return null;
  }
}

async function main() {
  console.log('='.repeat(60));
  console.log('🧪 TEST TOOL CLIENT - FETCH RESOURCES FROM SERVER');
  console.log('='.repeat(60));
  console.log();
  
  try {
    // Step 1: Login
    const auth = await login(USERNAME, PASSWORD);
    
    if (!auth) {
      console.log('\n❌ Không thể tiếp tục test do login thất bại');
      return;
    }
    
    console.log('-'.repeat(60));
    
    // Step 2: Get Resources
    const resources = await getResources(auth.userId, auth.token);
    
    if (resources) {
      console.log('\n' + '='.repeat(60));
      console.log('✅ TEST THÀNH CÔNG!');
      console.log('='.repeat(60));
      console.log('\n📝 Workflow cho tool thật:');
      console.log('   1. User nhập username/password trong tool');
      console.log('   2. Tool gọi /api/auth/login → lấy token + userId');
      console.log('   3. Tool gọi /api/resources/{userId} → lấy tất cả resources');
      console.log('   4. Tool load API keys, proxies, rotating keys vào memory');
      console.log('   5. Tool sẵn sàng xử lý batch!');
    } else {
      console.log('\n❌ TEST THẤT BẠI!');
    }
  } catch (error) {
    console.error('❌ Lỗi:', error.message);
    console.log('\n⚠️  Đảm bảo server đang chạy: npm run dev');
  }
}

main();

